src/nbl/ext/ImGui/ImGui.cpp

```diff
@@ -596,106 +924,289 @@ namespace nbl::ext::imgui
+										indirect->indexCount = pcmd->ElemCount;
+										indirect->instanceCount = 1u;
+										indirect->firstIndex = pcmd->IdxOffset + cmdListIndexObjectOffset;
+										indirect->vertexOffset = pcmd->VtxOffset + cmdListVertexObjectOffset;
+									}
+									else if (type == MDI::EBC_ELEMENT_STRUCTURES)
+									{
+										auto* element = reinterpret_cast<PerObjectData*>(data) + drawID;
+
+										const auto clipRectangle = clip.getClipRectangle(pcmd);
+										const auto scissor = clip.getScissor(clipRectangle);
+
+										auto packSnorm16 = [](float ndc) -> int16_t
+										{
+											return std::round<int16_t>(std::clamp(ndc, -1.0f, 1.0f) * 32767.0f); // TODO: ok encodePixels<asset::EF_R16_SNORM, double>(void* _pix, const double* _input) but iirc we have issues with our encode/decode utils
+										};
+
+										const auto vMin = trs.toNDC(core::vector2df_SIMD(scissor.offset.x, scissor.offset.y));
+										const auto vMax = trs.toNDC(core::vector2df_SIMD(scissor.offset.x + scissor.extent.width, scissor.offset.y + scissor.extent.height));
+
+										element->aabbMin.x = packSnorm16(vMin.x);
+										element->aabbMin.y = packSnorm16(vMin.y);
+										element->aabbMax.x = packSnorm16(vMax.x);
+										element->aabbMax.y = packSnorm16(vMax.y);
+										element->texId = pcmd->TextureId;
+									}
+
+									++drawID;
+								}
+
+								cmdListIndexObjectOffset += cmd_list->IdxBuffer.Size;
+								cmdListVertexObjectOffset += cmd_list->VtxBuffer.Size;
+							}
+
+							mdiBytesFilled[type] = true;
+							mdiOffsets[type] = globalBlockOffset;
+							return mdiParams.bytesToFill[type];
+						};
+
+						//! from biggest requests to smallest
+						uploadedSize += fillDrawBuffers.template operator() < MDI::EBC_VERTEX_BUFFERS > ();
+						uploadedSize += fillDrawBuffers.template operator() < MDI::EBC_INDEX_BUFFERS > ();
+						uploadedSize += fillIndirectStructures.template operator() < MDI::EBC_DRAW_INDIRECT_STRUCTURES > ();
+						uploadedSize += fillIndirectStructures.template operator() < MDI::EBC_ELEMENT_STRUCTURES > ();
+					}
+					streamingBuffer->multi_deallocate(STREAMING_ALLOCATION_COUNT, &requestState.offset, &requestState.multiAllocationSize, waitInfo); //! (*) block allocated, we just latch offsets deallocation to keep it alive as long as required
+				}
 			}
 
+			assert([&mdiOffsets]() -> bool
+			{
+				for (const auto& offset : mdiOffsets)
+					if (offset == INVALID_ADDRESS)
+						return false; // we should never hit this at this point
+
+				return true;
+			}()); // debug check only
+
+			auto mdiBuffer = smart_refctd_ptr<IGPUBuffer>(m_mdi.buffer->getBuffer());
+			const auto offset = mdiBuffer->getBoundMemory().offset;
 			{
 				const asset::SBufferBinding<const video::IGPUBuffer> binding =
 				{
-					.offset = 0,
-					.buffer = core::smart_refctd_ptr(indexBuffer)
+					.offset = mdiOffsets[MDI::EBC_INDEX_BUFFERS],
```

Comment:
dont bind at an offset, always bind at 0, then use the MDI parameters to control the index buffer offset

Reply:
discussed on discord - will give it another try